nengo/utils/tests/test_socket.py

```diff
@@ -0,0 +1,151 @@
+    m_send = nengo.Network(label='Send', seed=seed)
+    with m_send:
+        input = nengo.Node(output=lambda t: math.sin(10 * t))
+        socket_node = nengo.Node(size_in=1, output=udp_send.run)
+
+        nengo.Connection(input, socket_node, synapse=None)
+        p_s = nengo.Probe(input, synapse=None)
+    sim_send = Simulator(m_send)
+
+    # Model that receives data from previous model, feeds data back to itself,
+    # and sends that data out again
+    udp_both = sockets.UDPSocket(dest_port=54322, local_port=54321,
+                                 send_dim=1, recv_dim=1, timeout=1)
+    m_both = nengo.Network(label='Both', seed=seed)
+    with m_both:
+        socket_node = nengo.Node(size_in=1, output=udp_both.run)
+
+        nengo.Connection(socket_node, socket_node, synapse=0)
+        p_b = nengo.Probe(socket_node, synapse=None)
+    sim_both = Simulator(m_both)
+
+    # Model that receives data from previous model
+    udp_recv = sockets.UDPSocket(local_port=54322, recv_dim=1, timeout=1)
+    m_recv = nengo.Network(label='Recv', seed=seed)
+    with m_recv:
+        socket_node = nengo.Node(output=udp_recv.run)
+        p_r = nengo.Probe(socket_node, synapse=None)
+    sim_recv = Simulator(m_recv)
+
+    # Create thread lock to lock sending thread until receiving thread is ready
+    thread_lock = threading.Event()
+
+    # Create threads to run simulations
+    sim_thread_send = SimThread(sim_send, 0.5, thread_lock, True)
+    sim_thread_both = SimThread(sim_both, 0.5, thread_lock, True)
+    sim_thread_recv = SimThread(sim_recv, 0.5, thread_lock, False)
+    sim_thread_recv.start()
+    sim_thread_both.start()
+    sim_thread_send.start()
+    sim_thread_send.join()
+    sim_thread_both.join()
+    sim_thread_recv.join()
+
+    # Cleanup sockets
+    udp_send.close()
+    udp_both.close()
+    udp_recv.close()
+
+    # Do plots
+    plt.subplot(3, 1, 1)
+    plt.plot(sim_send.trange(), sim_send.data[p_s])
+    plt.title('1: Send node. Sends to 2.')
+    plt.subplot(3, 1, 2)
+    plt.plot(sim_both.trange(), sim_both.data[p_b])
+    plt.title('2: Send and recv node. Recvs from 1, sends to 3.')
+    plt.subplot(3, 1, 3)
+    plt.plot(sim_recv.trange(), sim_recv.data[p_r])
+    plt.title('3: Recv node. Recvs from 2.')
+
+    # Note: The socket communication delays information by 1 timestep in m_both
+    assert np.allclose(sim_send.data[p_s], sim_both.data[p_b],
+                       atol=0.0001, rtol=0.0001)
+    assert np.allclose(sim_both.data[p_b][:-1], sim_recv.data[p_r][1:],
+                       atol=0.0001, rtol=0.0001)
+
+
+def test_time_sync(Simulator, plt, seed, rng):
+    udp1 = sockets.UDPSocket(dest_port=54322, local_port=54321,
+                             send_dim=1, recv_dim=2, timeout=1)
+    m1 = nengo.Network(label='One', seed=seed)
+    with m1:
+        input = nengo.Node(output=lambda t: math.sin(10 * t))
+        socket_node = nengo.Node(size_in=1, output=udp1.run)
```

Comment:
I don't like having to pass `.run`. Can these be done as processes so that this isn't necessary? Or would that be confusing because with Processes it's typically fine to have more than one instance (and they're typically independent), whereas with sockets that's not necessarily the case?